### ISSUE TITLE
optimize function definitions for esbuild

### ISSUE THREAD

**Issue by borkdude** [squint-cljs/squint#511]
Currently functions are defined as:

```
var x = function () ...
```

Rather we could define functions with:

```
function x() { }
```

It seems esbuild won't even remove the prior notation if the function is unused.

**Comment by seancrowe** [comment]
Hi @borkdude,

Trying to be helpful here and say that I am not experiencing anything unexpected with esbuild with current notation.

## Scenario 1

If we create a file called `clojure.cljs` with this code
```clojure
(defn x [] ())
(defn z [] ())
(defn- y [] ()
```

If you compile with squint you get the file `clojure.mjs`
```js
import * as squint_core from 'squint-cljs/core.js';
var x = function () {
return squint_core.list();
};
var z = function () {
return squint_core.list();
};
var y = function () {
return squint_core.list();
};

export { x, z }
```

Then when you bundle with esbuild 
```bash
esbuild clojure.mjs --bundle --outdir=dist --platform=browser --tree-shaking=true
```

You will find in your final output file the functions `x` and `z`.

However, that is expected right, because we are `exporting { x, z }` and esbuild does not know if `x` or `z` will be used by another file.

If you delete say `x` from the export in your compiled `clojure.mjs` file like below
```js
import * as squint_core from 'squint-cljs/core.js';
var x = function () {
return squint_core.list();
};
var z = function () {
return squint_core.list();
};
var y = function () {
return squint_core.list();
};

export { z }
```
and run esbuild again, you will find that the function `x` is removed in the output.

## Scenario 2

If we make a `main.mjs` which looks like this:
```javascript
import { z } from "./clojure.mjs"

z()
```

And we go back and fix our `clojure.mjs` 😄 by compiling it with squint again so that the file exports `z` and `x` again.

Then we bundle esbuild `main.mjs`
```shell
esbuild main.mjs --bundle --outdir=dist --platform=browser --tree-shaking=true
```

We will find that our output only has `z`, and `x` was removed. That is expected because `main.mjs` is relying on `clojure.mjs` and only `z` is used

----
In summary, everything is working for me as expected using esbuild.

Moving to the notation change in theory would have other benfits. I have been told over my life that the black magic of JavaScript engines would optimize `function x () {}` over `var x = function () {}` and less overhead, but how true that is or to what extent, I never investigated.

See https://github.com/seancrowe/clojurescript-export-example for an example of the scenarios I mentioned above.

**Comment by borkdude** [comment]
Gotcha, thank you. Still I think `(defn foo [])` could maybe still compile to `function() foo {}` as that might be the more idiomatic way to write it in JS and some tools might expect you to write it like that.  Since the two ways of defining a function is semantically equivalent (both are mutable vars I believe) but only differ cosmetically, it might still be worth doing.
Thanks for explaining that esbuild still properly removes unused functions though, makes sense!

**Comment by kigiri** [comment]
It is also a common pattern to write functions with `const` now to prevent changing the name, ex:

```js
const add = (a, b) => a + b
```

**Comment by seancrowe** [comment]
JS functions can be written in three ways:
- Named: `function x () {}`
- Anonymous: `var x = function () {}`
- Arrow: `var x = () => {}`

## Equivalence
I agree it is not idiomatic to bind anonymous functions to variables (they were mainly used for callbacks pre-arrow functions), they are not semantically identical to named functions.

Named functions will be defined and initialized, hoisted, before the rest of the code is run, which is different than anonymous or arrow functions. Consider this Clojure code:

```clojure
(addTwo 5 3)
(defn addTwo [x y] (+ x y))
```

When compiled to JavaScript with Squint, it translates to:

```js
addTwo(5, 3);
var addTwo = function(x, y) {
  return (x) + (y);
};
```
If you run this JavaScript, it will error out saying "addTwo is not a function" because while the var addTwo is hoisted is left undefined.


Change `var addTwo` to `function addTwo`:

```js
addTwo(5, 3);
function addTwo(x, y) {
  return (x) + (y);
};
```

Run this JavaScript and it works because the named function `addTwo` is hoisted and defined. This behavior might confuse those familiar with Clojure, because in Clojure (to my knowledge) is top-down execution, not hoisting is done.

Hence, it's best to avoid named functions if you wish the output JavaScript to align with Clojure's behavior of top-down execution.

## Relevance of Idioms in Module Context
With Squint outputting modules, the choice between idiomatic and non-idiomatic JavaScript is less critical. This is because module exported functions are inherently read-only. The distinction between named, anonymous, or arrow functions, and the use of `var` vs. `const` or `let`, is moot. To the outside world, which is importing your functions, it is just a function that is similar to being declared by `const`.

For direct JavaScript manipulation or in cases where JavaScript is maintained post-compilation, these things would matter. So, it is not like you can completely ignore things, but if nobody is really looking at the JavaScript output, you'd be forgiven for not being idiomatic. 


## Suggestion
My take? Keep things as they are unless optimization issues are verified by using anonymous functions over named. Even then, you might want the behavior of the compiler to be configurable as by default I assume Squint's output JS should behave similar in execution order as the original Clojure code.

As for the idiomatic use of arrow functions, that could be an alternative. However, I personally know of a situation where `const` was replaced with `var` in a for loop and there was a measurable performance gain. Since JavaScript output is not meant to be maintained post-compilation, there is not a compelling reason to prefer `const` over `var`. Although, use cases may change and that too could be configurable in squint.edn.

**Comment by borkdude** [comment]
> As for the idiomatic use of arrow functions, that could be an alternative.

I don't think this is a good alternative since `function` and `=>` don't behave the same with respect to `this` and also you can't write generator functions using `=>`. This is why `function` is the default (to closely align with CLJS).

**Comment by borkdude** [comment]
But yeah, I think we can revisit this when there's a good reason to.
